jina/drivers/rank.py

```diff
@@ -54,15 +71,76 @@ def _apply_all(self, docs: 'jina_pb2.Document', context_doc: 'jina_pb2.Document'
 
         # np.uint32 uses 32 bits. np.float32 uses 23 bit mantissa, so integer greater than 2^23 will have their
         # least significant bits truncated.
+        if match_idx:
+            match_idx = np.array(match_idx, dtype=np.float64)
+
+            docs_scores = self.exec_fn(match_idx, query_chunk_meta, match_chunk_meta)
+            for doc_id, score in docs_scores:
+                r = context_doc.matches.add()
+                r.id = int(doc_id)
+                r.score.ref_id = context_doc.id  # label the score is computed against doc
+                r.score.value = score
+                r.score.op_name = exec.__class__.__name__
+
+
+class CollectMatches2DocRankDriver(BaseRankDriver):
+    """This Driver is intended to take a `document` with matches at a `given level depth > 0`, clear those matches and substitute
+    these matches by the documents at a lower depth level.
+    Input-Output ::
+        Input:
+        document: {level_depth: k}
+            |- matches: {level_depth: k}
+        Output:
+        document: {level_depth: k}
+            |- matches: {level_depth: k-1} (Sorted according to Ranker Executor)
+
+    Imagine a case where we are querying a system with text documents chunked by sentences. When we query the system,
+    we use sentences (chunks) to query it. So at some point we will have:
+    `query sentence (documents of level_depth 1):
+        matches: indexed sentences (documents of level depth 1)`
+    `
+    But in the output we want to have the full document that better matches the `sentence`.
+    `query sentence (documents of level_depth 1):
+        matches: indexed full documents (documents of level depth 0).
+    `
+    Using this Driver before querying a Binary Index with full binary document data can be very useful to implement a search system.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.recursion_order = 'post'
+
+    def _apply_all(self, context_doc: 'jina_pb2.Document', *args, **kwargs):
```

Comment:
this has an inconsistent signature as `_apply_all` of the parent class

Reply:
I did that on purpose because although not best practice if used with named parameters it would work. Since the matched documents would come from the context doc, I thought having a single point of entry of info would make the code less error prone and more self explainable.